# <font color=blue>Assignments for "Understanding The Relationship"</font>

To close out this lesson, you're going to do three assignments. For the first assignment, you'll write up a short answer to a question in a Gist file.  For the second two assignments, you'll do your work in Jupyter notebooks, and you should link to those notebooks in the same Gist file.

Please submit a single Gist file containing the answer to first assignment, plus links for second two.

## 1. Interpretion and signficance

Suppose that we would like to know how much families in the US are spending on recreation annually. We estimated the following model:

$$ expenditure = 873 + 0.0012annual\_income + 0.00002annual\_income^2 - 223.57have\_kids $$

*expenditure* is the annual spending on recreation in US dollars, *annual_income* is the annual income in US dollars and *have_kids* is a dummy variable indicating the families with children. Interpret the estimated coefficients. What additional statistics should be given in order to make sure that your interpretations make sense statistically. Write up your answer and save in a Gist. 

- The coefficient balancing the model is 873. This is also the bias value. While annual_income affected spending positively with a coefficient of 0.0012, annual_income2 also affected positively with a coefficient of 0.00002, and have_kids negatively affected it with a coefficient of 223.57.


## 2. Weather model

In this exercise, you'll work with the historical temperature data from the previous lesson. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

- First, load the dataset from the **weatherinszeged** table from Kaggle.
- Build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. Are the estimated coefficients statistically significant? Are the signs of the estimated coefficients in line with your previous expectations? Interpret the estimated coefficients. What are the relations between the target and the explanatory variables? 
- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for *humidity* and *windspeed* change? Interpret the estimated coefficients.

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas.api.types import is_numeric_dtype
from sklearn import metrics
import statsmodels.api as sm
import math
from sklearn import linear_model

df = pd.read_csv("C:/Users/Elif/data/weatherHistory.csv")

In [16]:

df["Temp_diff"] = df["Temperature (C)"] - df["Apparent Temperature (C)"]

Y = df["Temp_diff"]

X =df[["Humidity","Wind Speed (km/h)"]]

lrm = linear_model.LinearRegression()

# fit method estimates the coefficients using OLS
lrm.fit(X, Y)



X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              Temp_diff   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sat, 19 Sep 2020   Prob (F-statistic):               0.00
Time:                        20:43:11   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The estimated coefficients are statistically **significant**  because p-values< 0.05.

$$ TempatureDifference = -2.4381 + 3.0292Humidity + 0.1193Wind Speed (km/h) $$

In [18]:
df["Temp_diff"] = df["Temperature (C)"] - df["Apparent Temperature (C)"]

Y = df["Temp_diff"]
df['humwind'] = df['Humidity'] * df['Wind Speed (km/h)']
X =df[["Humidity","Wind Speed (km/h)","humwind"]]

lrm = linear_model.LinearRegression()

# fit method estimates the coefficients using OLS
lrm.fit(X, Y)



X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              Temp_diff   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sat, 19 Sep 2020   Prob (F-statistic):               0.00
Time:                        20:43:28   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
humwind               0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Creating new features slightly increased the value of the R-squared success parameter.

##  3. House prices model

In this exercise, you'll interpret your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

- Load the **houseprices** data from Kaggle.
- Run your house prices model again and interpret the results. Which features are statistically significant and which are not?
- Now, exclude the insignificant features from your model. Did anything change?
- Interpret the statistically significant coefficients by quantifying their relations with the house prices. Which features have more prominent effect on the house prices?
- Do the results sound reasonable to you? If not, try to explain the potential reasons.

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas.api.types import is_numeric_dtype
from sklearn import metrics
import statsmodels.api as sm
import math
from sklearn import linear_model

df = pd.read_csv("C:/Users/Elif/data/house_prices.csv")

In [6]:
df=df.drop(['PoolQC','MiscFeature','Fence','Alley'], axis=1)

In [9]:
def fix_missing(df, col, name):
    if is_numeric_dtype(col):
        df[name] = col.fillna(col.median())    
for n, c in df.items():
        fix_missing(df, c, n)

In [11]:
Y = df["SalePrice"]

X = df[["YearBuilt","TotalBsmtSF","1stFlrSF","FullBath","GarageArea",'GrLivArea','OverallQual']]
X = sm.add_constant(X)

results = sm.OLS(Y,X).fit()
display(results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.768
Model:                            OLS   Adj. R-squared:                  0.767
Method:                 Least Squares   F-statistic:                     688.2
Date:                Sat, 19 Sep 2020   Prob (F-statistic):               0.00
Time:                        19:45:28   Log-Likelihood:                -17476.
No. Observations:                1460   AIC:                         3.497e+04
Df Residuals:                    1452   BIC:                         3.501e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -8.112e+05   8.97e+04     -9.039      0.000   -9.87e+05   -6.35e+05
YearBuilt     372.5621     47.208      7.892      0.000     279.958     465.166
TotalBsmtSF    18.0092      4.313      4.176      0.000       9.549      26.469
1stFlrSF       14.1704      4.981      2.845      0.005       4.399      23.942
FullBath    -4481.4104   2656.489     -1.687      0.092   -9692.377     729.556
GarageArea     43.3915      6.225      6.971      0.000      31.181      55.602
GrLivArea      51.2896      3.141     16.331      0.000      45.129      57.450
OverallQual  2.148e+04   1154.905     18.596      0.000    1.92e+04    2.37e+04
==============================================================================
Omnibus:                      525.297   Durbin-Watson:                   1.990
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            65024.498
Skew:                          -0.612   Prob(JB):                         0.00
Kurtosis:                      35.671   Cond. No.                     2.71e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.71e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Features are statistically significant except FullBath. Let's build the model without the FullBath feature.

In [13]:
X_ = df[["YearBuilt","TotalBsmtSF",
                 "1stFlrSF","GarageArea",'GrLivArea','OverallQual']]
X_ = sm.add_constant(X_)

results = sm.OLS(Y,X_).fit()
display(results.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.768
Model:                            OLS   Adj. R-squared:                  0.767
Method:                 Least Squares   F-statistic:                     801.4
Date:                Sat, 19 Sep 2020   Prob (F-statistic):               0.00
Time:                        19:47:31   Log-Likelihood:                -17478.
No. Observations:                1460   AIC:                         3.497e+04
Df Residuals:                    1453   BIC:                         3.501e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -7.542e+05   8.32e+04     -9.066      0.000   -9.17e+05   -5.91e+05
YearBuilt     342.2151     43.674      7.836      0.000     256.544     427.886
TotalBsmtSF    18.8874      4.284      4.409      0.000      10.484      27.291
1stFlrSF       13.7397      4.978      2.760      0.006       3.975      23.504
GarageArea     43.8013      6.224      7.038      0.000      31.592      56.010
GrLivArea      48.6625      2.729     17.831      0.000      43.309      54.016
OverallQual  2.134e+04   1152.660     18.511      0.000    1.91e+04    2.36e+04
==============================================================================
Omnibus:                      500.669   Durbin-Watson:                   1.986
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            59739.408
Skew:                          -0.526   Prob(JB):                         0.00
Kurtosis:                      34.319   Cond. No.                     2.51e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.51e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

$$ SalePrice=7.542e+05+342.2151YearBuilt+18.8874TotalBsmtSF+13.73971stFlrSF+43.8013GarageArea+48.6625GrLivArea+2.134e+04OverallQua $$